File: tests/xcalablemp/NOT-IMPLEMENTED/async/F/reduction_async2.f90
! testp029.f
! reduction指示文(*)のテスト

      program main
      include 'xmp_lib.h'
      integer,parameter:: N=1000
!$xmp nodes p(*)
!$xmp template t(N,N,N)
!$xmp distribute t(*,*,block) onto p
      integer a(N), sa, ansa
      real*8  b(N), sb, ansb
      real*4  c(N), sc, ansc
      character(len=2) result
!$xmp align a(i) with t(*,*,i)
!$xmp align b(i) with t(*,*,i)
!$xmp align c(i) with t(*,*,i)

      sa = 1
      sb = 1.0
      sc = 1.0
!$xmp loop on t(:,:,i)
      do i=1, N
         if(mod(i,100) .eq. 0) then
            a(i) = 2
            b(i) = 2.0
            c(i) = 2.0
         else
            a(i) = 1
            b(i) = 1.0
            c(i) = 1.0
         endif
      enddo
      
!$xmp loop on t(:,:,i)
      do i=1, N
         sa = sa*a(i)
      enddo

!$xmp reduction(*:sa) async(1)

!$xmp loop on t(:,:,i)
      do i=1, N
         sb = sb*b(i)
      enddo
!$xmp wait_async(1)
      sa = sa*3

!$xmp reduction(*:sb) async(1)

!$xmp wait_async(1)
      sb = sb*1.5

!$xmp loop on t(:,:,i)
      do i=1, N
         sc = sc*c(i)
      enddo

!$xmp reduction(*:sc) async(1)

!$xmp wait_async(1)
      sc = sc*0.5

      ansa = 1024*3
      ansb = 1024.0*1.5
      ansc = 1024.0*0.5

      result = 'OK'
      if(  sa .ne. ansa .or.
     $     abs(sb-ansb) .gt. 0.00001 .or.
     $     abs(sc-ansc) .gt. 0.00001) then
         result = 'NG'
      endif

      print *, xmp_node_num(), 'testp029.f ', result

      end

      

File: tests/XMP/NOT-IMPLEMENTED/59/reduction_async2.c
/*testp029.c*/
/*reduction指示文(*)のテスト*/
#include<mpi.h>
#include<xmp.h>
#include<stdio.h>
#include<stdlib.h>   
static const int N=1000;
#pragma xmp nodes p(*)
#pragma xmp template t(0:N-1,0:N-1,0:N-1)
#pragma xmp distribute t(*,*,block) onto p 
int a[N],sa,ansa;
double b[N],sb,ansb;
float c[N],sc,ansc;
char *result;
int i;
#pragma xmp align a[i] with t(*,*,i)
#pragma xmp align b[i] with t(*,*,i)
#pragma xmp align c[i] with t(*,*,i)
int main(void){

   sa=1;
   sb=1.0;
   sc=1.0;
#pragma xmp loop on t(:,:,i)
   for(i=0;i<N;i++){
      if(i%100 == 0){
         a[i]=2;
         b[i]=2.0;
         c[i]=2.0;
      }else{
         a[i]=1;
         b[i]=1.0;
         c[i]=1.0;
      }
   }
#pragma xmp loop on t(:,:,i)
   for(i=0;i<N;i++){
      sa = sa*a[i];
   } 
#pragma xmp reduction(*:sa) async(1)

#pragma xmp loop on t(:,:,i)
   for(i=0;i<N;i++){
      sb = sb*b[i];
   }
#pragma xmp wait_async(1)
   sa = sa*3;

#pragma xmp reduction(*:sb) async(1)

#pragma xmp wait_async(1)
   sb = sb*1.5;

#pragma xmp loop on t(:,:,i)
   for(i=0;i<N;i++){
      sc = sc*c[i];
   }

#pragma xmp reduction(*:sc) async(1)
#pragma xmp wait_async(1)
   sc = sc*0.5;

   ansa=1024*3;
   ansb=1024.0*1.5;
   ansc=1024.0*0.5;

   result = "OK";
   if(sa != ansa||abs(sb-ansb) > 0.000001||abs(sc-ansc) > 0.000001){ 
      result = "NG";
   }
 
   printf("%d %s %s\n",xmp_node_num(),"testp029.c",result);
   return 0;
}
      
         
      
   
